### ISSUE TITLE
ExecuteInNewAppDomain asplodes with Serialization issue

### ISSUE THREAD

**Issue by shiftkey** [Squirrel/OldSquirrelForWindows#36]
Yes, I've dusted off MarkPad. Yes, I'm trying to ship three different updates in one long weekend. No, you cannot judge me :P 

Running from source on master so that I can step through the code - let me know if there's code that you know isn't ready...

I get to the UpdateManager.installPackageToAppDir call [here](https://github.com/github/Shimmer/blob/master/src/Shimmer.Client/UpdateManager.cs#L186) which then tries to run some code in a separate appdomain.

However, it asplodes.

Branch: https://github.com/shiftkey/DownmarkerWPF/tree/xpaulbettsx-shimmer-plus-markpad-equals-heart 

**Message:**

```
Type 'System.Reactive.Disposables.AnonymousDisposable' in assembly 'System.Reactive.Core, Version=2.0.20823.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' is not marked as serializable.
```

**Stack Trace:**

```
at Shimmer.Core.MethodRunner.Execute[TIn,TOut](TIn input, Func`2 method)
at Shimmer.Core.AppDomainHelper.<>c__DisplayClass1`2.<ExecuteInNewAppDomain>b__0(MethodRunner runner, TIn arg) in d:\Code\github\shiftkey\Shimmer\src\Shimmer.Core\AppDomainHelper.cs:line 11
at Shimmer.Core.AppDomainHelper.runInNewAppDomain[TIn,TOut](TIn input, Func`3 method) in d:\Code\github\shiftkey\Shimmer\src\Shimmer.Core\AppDomainHelper.cs:line 44
at Shimmer.Core.AppDomainHelper.ExecuteInNewAppDomain[TIn,TOut](TIn input, Func`2 method) in d:\Code\github\shiftkey\Shimmer\src\Shimmer.Core\AppDomainHelper.cs:line 11
at Shimmer.Client.UpdateManager.runPostInstallOnDirectory(String newAppDirectoryRoot, Boolean isFirstInstall, Version newCurrentVersion, IEnumerable`1 shortcutRequestsToIgnore) in d:\Code\github\shiftkey\Shimmer\src\Shimmer.Client\UpdateManager.cs:line 610
at Shimmer.Client.UpdateManager.runPostInstallAndCleanup(Version newCurrentVersion, Boolean isBootstrapping) in d:\Code\github\shiftkey\Shimmer\src\Shimmer.Client\UpdateManager.cs:line 431
at Shimmer.Client.UpdateManager.installPackageToAppDir(UpdateInfo updateInfo, ReleaseEntry release) in d:\Code\github\shiftkey\Shimmer\src\Shimmer.Client\UpdateManager.cs:line 419
at Shimmer.Client.UpdateManager.<>c__DisplayClass24.<>c__DisplayClass26.<applyReleases>b__1e() in d:\Code\github\shiftkey\Shimmer\src\Shimmer.Client\UpdateManager.cs:line 186
at System.Reactive.Linq.QueryLanguage.<>c__DisplayClasscb`1.<>c__DisplayClasscd.<ToAsync>b__ca()
```

Will do some more digging and update with more info, gonna pour myself a scotch first.

**Comment by shiftkey** [comment]
So I swapped out this code in UpdateManager.cs [here](https://github.com/github/Shimmer/blob/master/src/Shimmer.Client/UpdateManager.cs#L252)

```
var ret = Disposable.Create(() =>
{
    theLock.Dispose();
    updateLock = null;
});
```

With my own implementation (actually just the same code, but [Serializable]):

```
var ret = new MyAnonymousDisposable(() =>
{
    theLock.Dispose();
    updateLock = null;
});
```

which then gives me a different error (same stack trace):

Type 'Shimmer.Client.UpdateManager+<>c__DisplayClass30' in assembly 'Shimmer.Client, Version=0.6.0.0, Culture=neutral, PublicKeyToken=null' is not marked as serializable.

Which is this guy from ildasm.

![image](https://f.cloud.github.com/assets/359239/102368/40f93fde-691e-11e2-9247-94a2f7188f3d.png)

Puzzling...

**Comment by shiftkey** [comment]
Right, so I'm not going crazy - the sample application shows the exact same behaviour. 

Let me know whether these changes are ready for those insane enough to be building from source.

I'll pipe up if I can help out with it, brain is fried right now.

**Comment by anaisbetts** [comment]
I think some sort of upgrade killed this - Shimmer is still on my radar, I've just got too long of a TODO list

**Comment by shiftkey** [comment]
All good, let me test it out with the source version of 0.6.0.0-beta and see if its happening there as well.

**Comment by anaisbetts** [comment]
This is fixed in 065508582d460dd, unclear why it ever worked in previous versions
